<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>21 Sequences alignment | EHA Modeling &amp; Analytics Handbook</title>
  <meta name="description" content="21 Sequences alignment | EHA Modeling &amp; Analytics Handbook" />
  <meta name="generator" content="bookdown 0.37 and GitBook 2.6.7" />

  <meta property="og:title" content="21 Sequences alignment | EHA Modeling &amp; Analytics Handbook" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="ecohealthalliance/eha-ma-handbook" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="21 Sequences alignment | EHA Modeling &amp; Analytics Handbook" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.png" type="image/x-icon" />
<link rel="prev" href="20-assembly-of-high-throughput-data.html"/>
<link rel="next" href="22-data-sources.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<script src="libs/viz-1.8.2/viz.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/grViz-binding-1.0.10/grViz.js"></script>
<link href="https://fonts.googleapis.com/css?family=Lato" rel="stylesheet">


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="http://www.ecohealthalliance.org" target="_blank"><img src="assets/eha_logo.png" style="max-width: 100%"></a></li>
<li><a href="./" style="text-align:center"><strong>Modeling & Analytics Handbook</strong><a/></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="1" data-path="1-Contributing.html"><a href="1-Contributing.html"><i class="fa fa-check"></i><b>1</b> Contributing</a>
<ul>
<li class="chapter" data-level="1.1" data-path="1-Contributing.html"><a href="1-Contributing.html#anatomy-of-a-chapter"><i class="fa fa-check"></i><b>1.1</b> Anatomy of a chapter</a></li>
<li class="chapter" data-level="1.2" data-path="1-Contributing.html"><a href="1-Contributing.html#workflow-summary"><i class="fa fa-check"></i><b>1.2</b> Workflow Summary</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="1-Contributing.html"><a href="1-Contributing.html#for-quick-edits"><i class="fa fa-check"></i><b>1.2.1</b> For quick edits</a></li>
<li class="chapter" data-level="1.2.2" data-path="1-Contributing.html"><a href="1-Contributing.html#adding-a-chapter"><i class="fa fa-check"></i><b>1.2.2</b> Adding a chapter</a></li>
<li class="chapter" data-level="1.2.3" data-path="1-Contributing.html"><a href="1-Contributing.html#major-edits"><i class="fa fa-check"></i><b>1.2.3</b> Major edits</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="1-Contributing.html"><a href="1-Contributing.html#git-based-workflows"><i class="fa fa-check"></i><b>1.3</b> Git based workflows</a></li>
<li class="chapter" data-level="1.4" data-path="1-Contributing.html"><a href="1-Contributing.html#working-with-rmarkdown-in-the-bookdown-framework"><i class="fa fa-check"></i><b>1.4</b> Working with rmarkdown in the bookdown framework</a></li>
<li class="chapter" data-level="1.5" data-path="1-Contributing.html"><a href="1-Contributing.html#review"><i class="fa fa-check"></i><b>1.5</b> Review</a></li>
<li class="chapter" data-level="1.6" data-path="1-Contributing.html"><a href="1-Contributing.html#modifying-chapters"><i class="fa fa-check"></i><b>1.6</b> Modifying chapters</a></li>
<li class="chapter" data-level="1.7" data-path="1-Contributing.html"><a href="1-Contributing.html#additional-resources"><i class="fa fa-check"></i><b>1.7</b> Additional resources</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="2-quickstart-for-computing.html"><a href="2-quickstart-for-computing.html"><i class="fa fa-check"></i><b>2</b> Quickstart for computing</a>
<ul>
<li class="chapter" data-level="2.1" data-path="2-quickstart-for-computing.html"><a href="2-quickstart-for-computing.html#optional"><i class="fa fa-check"></i><b>2.1</b> Optional</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="3-projects.html"><a href="3-projects.html"><i class="fa fa-check"></i><b>3</b> Project Management</a>
<ul>
<li class="chapter" data-level="3.1" data-path="3-projects.html"><a href="3-projects.html#teams-and-work-cycles"><i class="fa fa-check"></i><b>3.1</b> Teams and Work Cycles</a></li>
<li class="chapter" data-level="3.2" data-path="3-projects.html"><a href="3-projects.html#setup-and-materials-organization"><i class="fa fa-check"></i><b>3.2</b> Setup and materials Organization</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="3-projects.html"><a href="3-projects.html#code-organization"><i class="fa fa-check"></i><b>3.2.1</b> Code organization</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="4-r.html"><a href="4-r.html"><i class="fa fa-check"></i><b>4</b> R and Reproducible Analysis</a>
<ul>
<li class="chapter" data-level="4.1" data-path="4-r.html"><a href="4-r.html#install"><i class="fa fa-check"></i><b>4.1</b> Install</a></li>
<li class="chapter" data-level="4.2" data-path="4-r.html"><a href="4-r.html#learn"><i class="fa fa-check"></i><b>4.2</b> Learn</a></li>
<li class="chapter" data-level="4.3" data-path="4-r.html"><a href="4-r.html#additional-resources-1"><i class="fa fa-check"></i><b>4.3</b> Additional Resources</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="5-eha-team-communication.html"><a href="5-eha-team-communication.html"><i class="fa fa-check"></i><b>5</b> EHA Team Communication</a>
<ul>
<li class="chapter" data-level="5.1" data-path="5-eha-team-communication.html"><a href="5-eha-team-communication.html#install-1"><i class="fa fa-check"></i><b>5.1</b> Install</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="6-documentation-and-outputs.html"><a href="6-documentation-and-outputs.html"><i class="fa fa-check"></i><b>6</b> Documentation and Outputs</a>
<ul>
<li class="chapter" data-level="6.1" data-path="6-documentation-and-outputs.html"><a href="6-documentation-and-outputs.html#learn-1"><i class="fa fa-check"></i><b>6.1</b> Learn</a></li>
<li class="chapter" data-level="6.2" data-path="6-documentation-and-outputs.html"><a href="6-documentation-and-outputs.html#install-2"><i class="fa fa-check"></i><b>6.2</b> Install</a></li>
<li class="chapter" data-level="6.3" data-path="6-documentation-and-outputs.html"><a href="6-documentation-and-outputs.html#library"><i class="fa fa-check"></i><b>6.3</b> Library</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="7-data-management.html"><a href="7-data-management.html"><i class="fa fa-check"></i><b>7</b> Data Management</a>
<ul>
<li class="chapter" data-level="7.1" data-path="7-data-management.html"><a href="7-data-management.html#data-management-plan"><i class="fa fa-check"></i><b>7.1</b> Data Management Plan</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="7-data-management.html"><a href="7-data-management.html#expectations-by-project-phase"><i class="fa fa-check"></i><b>7.1.1</b> Expectations by project phase</a></li>
<li class="chapter" data-level="7.1.2" data-path="7-data-management.html"><a href="7-data-management.html#using-dmptool-to-create-prepare-your-proposal-data-management-plan"><i class="fa fa-check"></i><b>7.1.2</b> Using DMPTool to create prepare your proposal data Management plan</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="7-data-management.html"><a href="7-data-management.html#notes-on-data-management"><i class="fa fa-check"></i><b>7.2</b> Notes on data management</a></li>
<li class="chapter" data-level="7.3" data-path="7-data-management.html"><a href="7-data-management.html#backups"><i class="fa fa-check"></i><b>7.3</b> Backups</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="7-data-management.html"><a href="7-data-management.html#what-do-we-envision-these-backups-being-used-for"><i class="fa fa-check"></i><b>7.3.1</b> What do we envision these backups being used for?</a></li>
<li class="chapter" data-level="7.3.2" data-path="7-data-management.html"><a href="7-data-management.html#what-counts-as-a-back-up"><i class="fa fa-check"></i><b>7.3.2</b> What counts as a back-up?</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="7-data-management.html"><a href="7-data-management.html#learn-2"><i class="fa fa-check"></i><b>7.4</b> Learn</a></li>
<li class="chapter" data-level="7.5" data-path="7-data-management.html"><a href="7-data-management.html#install-3"><i class="fa fa-check"></i><b>7.5</b> Install</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="8-airtable.html"><a href="8-airtable.html"><i class="fa fa-check"></i><b>8</b> Airtable</a>
<ul>
<li class="chapter" data-level="8.1" data-path="8-airtable.html"><a href="8-airtable.html#key-terms"><i class="fa fa-check"></i><b>8.1</b> Key Terms</a></li>
<li class="chapter" data-level="8.2" data-path="8-airtable.html"><a href="8-airtable.html#what-is-airtable"><i class="fa fa-check"></i><b>8.2</b> What is Airtable?</a></li>
<li class="chapter" data-level="8.3" data-path="8-airtable.html"><a href="8-airtable.html#when-should-i-use-airtable"><i class="fa fa-check"></i><b>8.3</b> When should I use Airtable?</a></li>
<li class="chapter" data-level="8.4" data-path="8-airtable.html"><a href="8-airtable.html#security-access-control-and-data-durability"><i class="fa fa-check"></i><b>8.4</b> Security, Access Control, and Data Durability</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="8-airtable.html"><a href="8-airtable.html#user--and-administrator-security-features"><i class="fa fa-check"></i><b>8.4.1</b> User- and Administrator Security Features</a></li>
<li class="chapter" data-level="8.4.2" data-path="8-airtable.html"><a href="8-airtable.html#share-links-and-interfaces"><i class="fa fa-check"></i><b>8.4.2</b> Share Links and Interfaces</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="8-airtable.html"><a href="8-airtable.html#data-in-airtable"><i class="fa fa-check"></i><b>8.5</b> Data in Airtable</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="8-airtable.html"><a href="8-airtable.html#workspaces-bases-tables-fields-records"><i class="fa fa-check"></i><b>8.5.1</b> Workspaces, Bases, Tables, Fields, Records</a></li>
<li class="chapter" data-level="8.5.2" data-path="8-airtable.html"><a href="8-airtable.html#views"><i class="fa fa-check"></i><b>8.5.2</b> Views</a></li>
<li class="chapter" data-level="8.5.3" data-path="8-airtable.html"><a href="8-airtable.html#internal-backups-record-history-base-snapshots-and-duplication"><i class="fa fa-check"></i><b>8.5.3</b> Internal Backups: Record History, Base Snapshots and Duplication</a></li>
<li class="chapter" data-level="8.5.4" data-path="8-airtable.html"><a href="8-airtable.html#importing-datad"><i class="fa fa-check"></i><b>8.5.4</b> Importing datad</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="8-airtable.html"><a href="8-airtable.html#base-design"><i class="fa fa-check"></i><b>8.6</b> Base Design</a></li>
<li class="chapter" data-level="8.7" data-path="8-airtable.html"><a href="8-airtable.html#automating-airtable"><i class="fa fa-check"></i><b>8.7</b> Automating Airtable</a>
<ul>
<li class="chapter" data-level="8.7.1" data-path="8-airtable.html"><a href="8-airtable.html#automations-with-drag-and-drop-programing-pro-and-above"><i class="fa fa-check"></i><b>8.7.1</b> Automations with Drag and Drop Programing (pro and above)</a></li>
<li class="chapter" data-level="8.7.2" data-path="8-airtable.html"><a href="8-airtable.html#scripting-and-blocks-pro-plan-and-above"><i class="fa fa-check"></i><b>8.7.2</b> Scripting and Blocks (pro plan and above)</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="8-airtable.html"><a href="8-airtable.html#using-the-rest-api"><i class="fa fa-check"></i><b>8.8</b> Using the REST API</a>
<ul>
<li class="chapter" data-level="8.8.1" data-path="8-airtable.html"><a href="8-airtable.html#scoped-tokens"><i class="fa fa-check"></i><b>8.8.1</b> Scoped Tokens</a></li>
<li class="chapter" data-level="8.8.2" data-path="8-airtable.html"><a href="8-airtable.html#airtable-and-r"><i class="fa fa-check"></i><b>8.8.2</b> Airtable and R</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="8-airtable.html"><a href="8-airtable.html#data-management-1"><i class="fa fa-check"></i><b>8.9</b> Data Management</a>
<ul>
<li class="chapter" data-level="8.9.1" data-path="8-airtable.html"><a href="8-airtable.html#metadata"><i class="fa fa-check"></i><b>8.9.1</b> Metadata</a></li>
<li class="chapter" data-level="8.9.2" data-path="8-airtable.html"><a href="8-airtable.html#external-backups"><i class="fa fa-check"></i><b>8.9.2</b> External Backups</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="9-versioning.html"><a href="9-versioning.html"><i class="fa fa-check"></i><b>9</b> Version Control, Git and Github</a>
<ul>
<li class="chapter" data-level="9.1" data-path="9-versioning.html"><a href="9-versioning.html#learn-3"><i class="fa fa-check"></i><b>9.1</b> Learn</a></li>
<li class="chapter" data-level="9.2" data-path="9-versioning.html"><a href="9-versioning.html#install-4"><i class="fa fa-check"></i><b>9.2</b> Install</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="9-versioning.html"><a href="9-versioning.html#creating-tokens"><i class="fa fa-check"></i><b>9.2.1</b> Creating Tokens</a></li>
<li class="chapter" data-level="9.2.2" data-path="9-versioning.html"><a href="9-versioning.html#regenerating-tokens"><i class="fa fa-check"></i><b>9.2.2</b> Regenerating Tokens</a></li>
<li class="chapter" data-level="9.2.3" data-path="9-versioning.html"><a href="9-versioning.html#useful-references"><i class="fa fa-check"></i><b>9.2.3</b> Useful References</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="10-reviewing-analyses-and-code.html"><a href="10-reviewing-analyses-and-code.html"><i class="fa fa-check"></i><b>10</b> Reviewing Analyses and Code</a>
<ul>
<li class="chapter" data-level="10.1" data-path="10-reviewing-analyses-and-code.html"><a href="10-reviewing-analyses-and-code.html#learn-4"><i class="fa fa-check"></i><b>10.1</b> Learn</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="11-testing.html"><a href="11-testing.html"><i class="fa fa-check"></i><b>11</b> Testing</a>
<ul>
<li class="chapter" data-level="11.1" data-path="11-testing.html"><a href="11-testing.html#learn-5"><i class="fa fa-check"></i><b>11.1</b> Learn</a></li>
<li class="chapter" data-level="11.2" data-path="11-testing.html"><a href="11-testing.html#install-5"><i class="fa fa-check"></i><b>11.2</b> Install</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="12-high-performance-servers.html"><a href="12-high-performance-servers.html"><i class="fa fa-check"></i><b>12</b> High-Performance Servers</a></li>
<li class="chapter" data-level="13" data-path="13-odk.html"><a href="13-odk.html"><i class="fa fa-check"></i><b>13</b> Open Data Kit</a>
<ul>
<li class="chapter" data-level="13.1" data-path="13-odk.html"><a href="13-odk.html#central"><i class="fa fa-check"></i><b>13.1</b> ODK Central</a>
<ul>
<li class="chapter" data-level="13.1.1" data-path="13-odk.html"><a href="13-odk.html#getting-an-eha-odk-central-account"><i class="fa fa-check"></i><b>13.1.1</b> Getting an EHA ODK Central account</a></li>
<li class="chapter" data-level="13.1.2" data-path="13-odk.html"><a href="13-odk.html#using-odk-central"><i class="fa fa-check"></i><b>13.1.2</b> Using ODK Central</a></li>
<li class="chapter" data-level="13.1.3" data-path="13-odk.html"><a href="13-odk.html#setting-up-a-project"><i class="fa fa-check"></i><b>13.1.3</b> Setting up a project</a></li>
<li class="chapter" data-level="13.1.4" data-path="13-odk.html"><a href="13-odk.html#monitoring-the-progress-of-projects"><i class="fa fa-check"></i><b>13.1.4</b> Monitoring the progress of projects</a></li>
<li class="chapter" data-level="13.1.5" data-path="13-odk.html"><a href="13-odk.html#retrieving-data"><i class="fa fa-check"></i><b>13.1.5</b> Retrieving Data</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="13-odk.html"><a href="13-odk.html#collect"><i class="fa fa-check"></i><b>13.2</b> ODK Collect</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="13-odk.html"><a href="13-odk.html#installing-odk-collect"><i class="fa fa-check"></i><b>13.2.1</b> Installing ODK Collect</a></li>
<li class="chapter" data-level="13.2.2" data-path="13-odk.html"><a href="13-odk.html#connecting-to-a-server"><i class="fa fa-check"></i><b>13.2.2</b> Connecting to a server</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="13-odk.html"><a href="13-odk.html#other-useful-odk-collect-configurations"><i class="fa fa-check"></i><b>13.3</b> Other useful ODK Collect configurations</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="13-odk.html"><a href="13-odk.html#language-settings"><i class="fa fa-check"></i><b>13.3.1</b> Language settings</a></li>
<li class="chapter" data-level="13.3.2" data-path="13-odk.html"><a href="13-odk.html#localizing-odk-collect"><i class="fa fa-check"></i><b>13.3.2</b> Localizing ODK Collect</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="13-odk.html"><a href="13-odk.html#xlsform"><i class="fa fa-check"></i><b>13.4</b> XLSForm</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="14-cloud-computing-services.html"><a href="14-cloud-computing-services.html"><i class="fa fa-check"></i><b>14</b> Cloud Computing Services</a>
<ul>
<li class="chapter" data-level="14.1" data-path="14-cloud-computing-services.html"><a href="14-cloud-computing-services.html#setting-up-amazon-credentials"><i class="fa fa-check"></i><b>14.1</b> Setting up Amazon Credentials</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="14-cloud-computing-services.html"><a href="14-cloud-computing-services.html#generate-an-access-key"><i class="fa fa-check"></i><b>14.1.1</b> Generate an access key</a></li>
<li class="chapter" data-level="14.1.2" data-path="14-cloud-computing-services.html"><a href="14-cloud-computing-services.html#create-a-credentials-file"><i class="fa fa-check"></i><b>14.1.2</b> Create a credentials file</a></li>
<li class="chapter" data-level="14.1.3" data-path="14-cloud-computing-services.html"><a href="14-cloud-computing-services.html#test-your-key"><i class="fa fa-check"></i><b>14.1.3</b> Test your Key</a></li>
<li class="chapter" data-level="14.1.4" data-path="14-cloud-computing-services.html"><a href="14-cloud-computing-services.html#alternate-credentialing-method"><i class="fa fa-check"></i><b>14.1.4</b> Alternate credentialing method</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="14-cloud-computing-services.html"><a href="14-cloud-computing-services.html#using-amazon-s3-storage-from-r"><i class="fa fa-check"></i><b>14.2</b> Using Amazon S3 Storage from R</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="14-cloud-computing-services.html"><a href="14-cloud-computing-services.html#save-objects-from-s3-to-your-machine"><i class="fa fa-check"></i><b>14.2.1</b> Save objects from S3 to your machine</a></li>
<li class="chapter" data-level="14.2.2" data-path="14-cloud-computing-services.html"><a href="14-cloud-computing-services.html#upload-objects-from-your-machine-to-s3"><i class="fa fa-check"></i><b>14.2.2</b> Upload objects from your machine to S3</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html"><i class="fa fa-check"></i><b>15</b> Google Authorization and R</a>
<ul>
<li class="chapter" data-level="15.1" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#the-basic-overview"><i class="fa fa-check"></i><b>15.1</b> The basic overview:</a></li>
<li class="chapter" data-level="15.2" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#key-terms-1"><i class="fa fa-check"></i><b>15.2</b> Key Terms</a></li>
<li class="chapter" data-level="15.3" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#before-we-start"><i class="fa fa-check"></i><b>15.3</b> Before we start:</a></li>
<li class="chapter" data-level="15.4" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#setup-encryption-tools-on-your-machine"><i class="fa fa-check"></i><b>15.4</b> Setup encryption tools on your machine</a></li>
<li class="chapter" data-level="15.5" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#enable-git-crypt-on-your-repository"><i class="fa fa-check"></i><b>15.5</b> Enable git-crypt on your repository</a></li>
<li class="chapter" data-level="15.6" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#setting-up-non-interactive-authentication-for-google-sheets"><i class="fa fa-check"></i><b>15.6</b> Setting up non-interactive authentication for Google sheets</a>
<ul>
<li class="chapter" data-level="15.6.1" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#create-a-google-cloud-platform-project"><i class="fa fa-check"></i><b>15.6.1</b> Create a Google cloud platform project</a></li>
<li class="chapter" data-level="15.6.2" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#enable-apis"><i class="fa fa-check"></i><b>15.6.2</b> Enable APIs</a></li>
<li class="chapter" data-level="15.6.3" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#create-a-service-account"><i class="fa fa-check"></i><b>15.6.3</b> Create a Service Account</a></li>
<li class="chapter" data-level="15.6.4" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#create-a-key-for-your-service-account"><i class="fa fa-check"></i><b>15.6.4</b> Create a Key for your service account</a></li>
<li class="chapter" data-level="15.6.5" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#share-the-google-drive-resources-of-interest-with-the-service-account"><i class="fa fa-check"></i><b>15.6.5</b> Share the Google drive resources of interest with the service account</a></li>
</ul></li>
<li class="chapter" data-level="15.7" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#general-approach-to-securely-managing-keys"><i class="fa fa-check"></i><b>15.7</b> General approach to securely managing keys</a>
<ul>
<li class="chapter" data-level="15.7.1" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#provide-a-service-account-key-for-projects"><i class="fa fa-check"></i><b>15.7.1</b> Provide a service account key for projects</a></li>
</ul></li>
<li class="chapter" data-level="15.8" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#additional-resources-2"><i class="fa fa-check"></i><b>15.8</b> Additional Resources</a>
<ul>
<li class="chapter" data-level="15.8.1" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#google-drive"><i class="fa fa-check"></i><b>15.8.1</b> Google Drive</a></li>
<li class="chapter" data-level="15.8.2" data-path="15-google-authorization-and-r.html"><a href="15-google-authorization-and-r.html#google-sheets"><i class="fa fa-check"></i><b>15.8.2</b> Google Sheets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="16-encryption.html"><a href="16-encryption.html"><i class="fa fa-check"></i><b>16</b> Encryption</a>
<ul>
<li class="chapter" data-level="16.1" data-path="16-encryption.html"><a href="16-encryption.html#set-up-keybase"><i class="fa fa-check"></i><b>16.1</b> Set up Keybase</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="16-encryption.html"><a href="16-encryption.html#installing-keybase-on-linux"><i class="fa fa-check"></i><b>16.1.1</b> Installing Keybase on Linux</a></li>
<li class="chapter" data-level="16.1.2" data-path="16-encryption.html"><a href="16-encryption.html#installing-keybase-on-macos-and-windows"><i class="fa fa-check"></i><b>16.1.2</b> Installing Keybase on macOS and Windows</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="16-encryption.html"><a href="16-encryption.html#install-gpg-and-git-crypt"><i class="fa fa-check"></i><b>16.2</b> Install <code>gpg</code> and <code>git-crypt</code></a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="16-encryption.html"><a href="16-encryption.html#install-gpg-and-git-crypt-on-macos"><i class="fa fa-check"></i><b>16.2.1</b> Install <code>gpg</code> and <code>git-crypt</code> on macOS</a></li>
<li class="chapter" data-level="16.2.2" data-path="16-encryption.html"><a href="16-encryption.html#install-gpg-and-git-crypt-on-windows"><i class="fa fa-check"></i><b>16.2.2</b> Install <code>gpg</code> and <code>git-crypt</code> on Windows</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="16-encryption.html"><a href="16-encryption.html#create-your-keybase-keys"><i class="fa fa-check"></i><b>16.3</b> Create your Keybase keys</a></li>
<li class="chapter" data-level="16.4" data-path="16-encryption.html"><a href="16-encryption.html#import-your-keys-to-your-local-keychain"><i class="fa fa-check"></i><b>16.4</b> Import your keys to your local keychain</a></li>
<li class="chapter" data-level="16.5" data-path="16-encryption.html"><a href="16-encryption.html#configure-gpg"><i class="fa fa-check"></i><b>16.5</b> Configure gpg</a>
<ul>
<li class="chapter" data-level="16.5.1" data-path="16-encryption.html"><a href="16-encryption.html#configure-gpg-on-macos-and-linux"><i class="fa fa-check"></i><b>16.5.1</b> Configure gpg on macOS (and Linux)</a></li>
<li class="chapter" data-level="16.5.2" data-path="16-encryption.html"><a href="16-encryption.html#configure-gpg-on-windows"><i class="fa fa-check"></i><b>16.5.2</b> Configure gpg on Windows</a></li>
</ul></li>
<li class="chapter" data-level="16.6" data-path="16-encryption.html"><a href="16-encryption.html#backing-up-and-recovering-gpg-keys"><i class="fa fa-check"></i><b>16.6</b> Backing up and Recovering GPG Keys</a>
<ul>
<li class="chapter" data-level="16.6.1" data-path="16-encryption.html"><a href="16-encryption.html#how-do-i-save-a-key"><i class="fa fa-check"></i><b>16.6.1</b> How do I save a key?</a></li>
<li class="chapter" data-level="16.6.2" data-path="16-encryption.html"><a href="16-encryption.html#how-do-i-recover-my-gpg-key-from-a-file"><i class="fa fa-check"></i><b>16.6.2</b> How do I recover my gpg key from a file?</a></li>
</ul></li>
<li class="chapter" data-level="16.7" data-path="16-encryption.html"><a href="16-encryption.html#use-git-crypt-to-unlock-a-repository"><i class="fa fa-check"></i><b>16.7</b> Use git-crypt to unlock a repository</a>
<ul>
<li class="chapter" data-level="16.7.1" data-path="16-encryption.html"><a href="16-encryption.html#troubleshooting"><i class="fa fa-check"></i><b>16.7.1</b> Troubleshooting</a></li>
</ul></li>
<li class="chapter" data-level="16.8" data-path="16-encryption.html"><a href="16-encryption.html#managing-which-users-can-decrypt-files-in-a-repository"><i class="fa fa-check"></i><b>16.8</b> Managing which users can decrypt files in a repository</a>
<ul>
<li class="chapter" data-level="16.8.1" data-path="16-encryption.html"><a href="16-encryption.html#set-up-encryption-for-a-repo-that-did-not-previously-use-git-crypt"><i class="fa fa-check"></i><b>16.8.1</b> Set up encryption for a repo that did not previously use git-crypt</a></li>
<li class="chapter" data-level="16.8.2" data-path="16-encryption.html"><a href="16-encryption.html#allow-contributors-access-to-encrypted-files"><i class="fa fa-check"></i><b>16.8.2</b> Allow contributors access to encrypted files</a></li>
<li class="chapter" data-level="16.8.3" data-path="16-encryption.html"><a href="16-encryption.html#import-your-keys-to-the-eha-server"><i class="fa fa-check"></i><b>16.8.3</b> Import your keys to the EHA server</a></li>
</ul></li>
<li class="chapter" data-level="16.9" data-path="16-encryption.html"><a href="16-encryption.html#use-a-symmetric-key-for-automated-processes"><i class="fa fa-check"></i><b>16.9</b> Use a symmetric key for automated processes</a></li>
<li class="chapter" data-level="16.10" data-path="16-encryption.html"><a href="16-encryption.html#removing-sensitive-files-from-git-history"><i class="fa fa-check"></i><b>16.10</b> Removing sensitive files from git history</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="17-dependencies.html"><a href="17-dependencies.html"><i class="fa fa-check"></i><b>17</b> Dependency Management</a>
<ul>
<li class="chapter" data-level="17.1" data-path="17-dependencies.html"><a href="17-dependencies.html#gcc-and-mac-silicon"><i class="fa fa-check"></i><b>17.1</b> GCC and mac silicon</a></li>
<li class="chapter" data-level="17.2" data-path="17-dependencies.html"><a href="17-dependencies.html#section"><i class="fa fa-check"></i><b>17.2</b> </a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="18-statistical-methods.html"><a href="18-statistical-methods.html"><i class="fa fa-check"></i><b>18</b> Statistical Methods</a>
<ul>
<li class="chapter" data-level="18.1" data-path="18-statistical-methods.html"><a href="18-statistical-methods.html#multivariate-regression"><i class="fa fa-check"></i><b>18.1</b> Multivariate Regression</a></li>
<li class="chapter" data-level="18.2" data-path="18-statistical-methods.html"><a href="18-statistical-methods.html#networks"><i class="fa fa-check"></i><b>18.2</b> Networks</a></li>
<li class="chapter" data-level="18.3" data-path="18-statistical-methods.html"><a href="18-statistical-methods.html#bioinformatics-and-sequence-analysis"><i class="fa fa-check"></i><b>18.3</b> Bioinformatics and Sequence Analysis</a></li>
<li class="chapter" data-level="18.4" data-path="18-statistical-methods.html"><a href="18-statistical-methods.html#species-distribution-modeling"><i class="fa fa-check"></i><b>18.4</b> Species Distribution Modeling</a></li>
<li class="chapter" data-level="18.5" data-path="18-statistical-methods.html"><a href="18-statistical-methods.html#epidemic-simulation-and-fitting"><i class="fa fa-check"></i><b>18.5</b> Epidemic Simulation and Fitting</a></li>
<li class="chapter" data-level="18.6" data-path="18-statistical-methods.html"><a href="18-statistical-methods.html#phylogenetics"><i class="fa fa-check"></i><b>18.6</b> Phylogenetics</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="19-quality-control.html"><a href="19-quality-control.html"><i class="fa fa-check"></i><b>19</b> Quality control</a>
<ul>
<li class="chapter" data-level="19.0.1" data-path="19-quality-control.html"><a href="19-quality-control.html#what-is-quality-control"><i class="fa fa-check"></i><b>19.0.1</b> What is quality control?</a></li>
<li class="chapter" data-level="19.0.2" data-path="19-quality-control.html"><a href="19-quality-control.html#why-is-a-quality-control-step-necessary-in-ngs-data"><i class="fa fa-check"></i><b>19.0.2</b> Why is a quality control step necessary in NGS data?</a></li>
<li class="chapter" data-level="19.0.3" data-path="19-quality-control.html"><a href="19-quality-control.html#why-is-a-quality-control-step-important-in-the-processing-of-ngs-data"><i class="fa fa-check"></i><b>19.0.3</b> Why is a quality control step important in the processing of NGS data?</a></li>
<li class="chapter" data-level="19.0.4" data-path="19-quality-control.html"><a href="19-quality-control.html#how-does-the-trimming-process-work"><i class="fa fa-check"></i><b>19.0.4</b> How does the trimming process work?</a></li>
<li class="chapter" data-level="19.0.5" data-path="19-quality-control.html"><a href="19-quality-control.html#quality-control-with-trimmomatic"><i class="fa fa-check"></i><b>19.0.5</b> Quality control with trimmomatic</a></li>
<li class="chapter" data-level="19.0.6" data-path="19-quality-control.html"><a href="19-quality-control.html#visualization-of-the-quality-of-the-reads"><i class="fa fa-check"></i><b>19.0.6</b> Visualization of the quality of the reads</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="20-assembly-of-high-throughput-data.html"><a href="20-assembly-of-high-throughput-data.html"><i class="fa fa-check"></i><b>20</b> Assembly of high-throughput data</a>
<ul>
<li class="chapter" data-level="20.0.1" data-path="20-assembly-of-high-throughput-data.html"><a href="20-assembly-of-high-throughput-data.html#the-problem"><i class="fa fa-check"></i><b>20.0.1</b> The problem</a></li>
<li class="chapter" data-level="20.0.2" data-path="20-assembly-of-high-throughput-data.html"><a href="20-assembly-of-high-throughput-data.html#graphs"><i class="fa fa-check"></i><b>20.0.2</b> Graphs</a></li>
<li class="chapter" data-level="20.0.3" data-path="20-assembly-of-high-throughput-data.html"><a href="20-assembly-of-high-throughput-data.html#bruijn-graph"><i class="fa fa-check"></i><b>20.0.3</b> Bruijn graph</a></li>
<li class="chapter" data-level="20.0.4" data-path="20-assembly-of-high-throughput-data.html"><a href="20-assembly-of-high-throughput-data.html#contigs"><i class="fa fa-check"></i><b>20.0.4</b> Contigs</a></li>
<li class="chapter" data-level="20.0.5" data-path="20-assembly-of-high-throughput-data.html"><a href="20-assembly-of-high-throughput-data.html#a-formal-presentation-of-de-bruijn-graphs"><i class="fa fa-check"></i><b>20.0.5</b> A formal presentation of de Bruijn graphs</a></li>
<li class="chapter" data-level="20.0.6" data-path="20-assembly-of-high-throughput-data.html"><a href="20-assembly-of-high-throughput-data.html#megahit"><i class="fa fa-check"></i><b>20.0.6</b> MEGAHIT</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html"><i class="fa fa-check"></i><b>21</b> Sequences alignment</a>
<ul>
<li class="chapter" data-level="21.1" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#heredity"><i class="fa fa-check"></i><b>21.1</b> Heredity</a></li>
<li class="chapter" data-level="21.2" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#orthologs-and-paralogs"><i class="fa fa-check"></i><b>21.2</b> Orthologs and paralogs</a></li>
<li class="chapter" data-level="21.3" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#blast"><i class="fa fa-check"></i><b>21.3</b> BLAST</a>
<ul>
<li class="chapter" data-level="21.3.1" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#conda-installation"><i class="fa fa-check"></i><b>21.3.1</b> Conda Installation</a></li>
</ul></li>
<li class="chapter" data-level="21.4" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#install-blast-with-conda"><i class="fa fa-check"></i><b>21.4</b> Install BLAST+ with Conda</a></li>
<li class="chapter" data-level="21.5" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#install-database-reference-sequences"><i class="fa fa-check"></i><b>21.5</b> Install database reference sequences</a>
<ul>
<li class="chapter" data-level="21.5.1" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#there-are-different-ways-to-install-the-pre-formatted-blast-databases"><i class="fa fa-check"></i><b>21.5.1</b> There are different ways to install the pre-formatted BLAST databases:</a></li>
</ul></li>
<li class="chapter" data-level="21.6" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#pairwise-sequence-alignment"><i class="fa fa-check"></i><b>21.6</b> Pairwise sequence alignment</a>
<ul>
<li class="chapter" data-level="21.6.1" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#blastn"><i class="fa fa-check"></i><b>21.6.1</b> BLASTn</a></li>
<li class="chapter" data-level="21.6.2" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#blastp"><i class="fa fa-check"></i><b>21.6.2</b> BLASTp</a></li>
<li class="chapter" data-level="21.6.3" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#configuration-files-and-environmental-variables"><i class="fa fa-check"></i><b>21.6.3</b> Configuration files and environmental variables</a></li>
</ul></li>
<li class="chapter" data-level="21.7" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#blast-with-high-throughput-data"><i class="fa fa-check"></i><b>21.7</b> BLAST with high-throughput data</a>
<ul>
<li class="chapter" data-level="21.7.1" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#installation-2"><i class="fa fa-check"></i><b>21.7.1</b> Installation</a></li>
<li class="chapter" data-level="21.7.2" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#add-sequence-database"><i class="fa fa-check"></i><b>21.7.2</b> Add sequence database</a></li>
<li class="chapter" data-level="21.7.3" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#add-the-job-to-the-queue"><i class="fa fa-check"></i><b>21.7.3</b> Add the job to the queue</a></li>
<li class="chapter" data-level="21.7.4" data-path="21-seqs_aln.html"><a href="21-seqs_aln.html#run-the-job"><i class="fa fa-check"></i><b>21.7.4</b> Run the job</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="22" data-path="22-data-sources.html"><a href="22-data-sources.html"><i class="fa fa-check"></i><b>22</b> Data Sources</a></li>
<li class="chapter" data-level="23" data-path="23-references.html"><a href="23-references.html"><i class="fa fa-check"></i><b>23</b> References</a>
<ul>
<li class="chapter" data-level="23.1" data-path="23-references.html"><a href="23-references.html#converting-paperpile-citations-in-google-docs-to-endnote-citations-in-ms-word"><i class="fa fa-check"></i><b>23.1</b> Converting Paperpile Citations in Google Docs to Endnote Citations in MS Word</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html"><i class="fa fa-check"></i><b>24</b> Training Resources and Plans</a>
<ul>
<li class="chapter" data-level="24.1" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#training-plan-components"><i class="fa fa-check"></i><b>24.1</b> Training plan components</a></li>
<li class="chapter" data-level="24.2" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#training-plans"><i class="fa fa-check"></i><b>24.2</b> Training Plans</a>
<ul>
<li class="chapter" data-level="24.2.1" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#introductory-programming-materials"><i class="fa fa-check"></i><b>24.2.1</b> Introductory Programming materials</a></li>
<li class="chapter" data-level="24.2.2" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#better-managing-data"><i class="fa fa-check"></i><b>24.2.2</b> Better Managing Data</a></li>
<li class="chapter" data-level="24.2.3" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#version-control-and-git"><i class="fa fa-check"></i><b>24.2.3</b> Version Control and Git</a></li>
<li class="chapter" data-level="24.2.4" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#reproducible-reporting"><i class="fa fa-check"></i><b>24.2.4</b> Reproducible reporting</a></li>
<li class="chapter" data-level="24.2.5" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#improving-your-statistical-fundamentals"><i class="fa fa-check"></i><b>24.2.5</b> Improving Your Statistical Fundamentals</a></li>
<li class="chapter" data-level="24.2.6" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#improving-your-data-visualization"><i class="fa fa-check"></i><b>24.2.6</b> Improving your data visualization</a></li>
<li class="chapter" data-level="24.2.7" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#map-making-and-geospatial-analysis-in-r"><i class="fa fa-check"></i><b>24.2.7</b> Map-making and geospatial analysis in R</a></li>
<li class="chapter" data-level="24.2.8" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#bioinformatics"><i class="fa fa-check"></i><b>24.2.8</b> Bioinformatics</a></li>
</ul></li>
<li class="chapter" data-level="24.3" data-path="24-training-resources-and-plans.html"><a href="24-training-resources-and-plans.html#metagenomics"><i class="fa fa-check"></i><b>24.3</b> Metagenomics</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="25-help.html"><a href="25-help.html"><i class="fa fa-check"></i><b>25</b> Getting Help</a>
<ul>
<li class="chapter" data-level="25.1" data-path="25-help.html"><a href="25-help.html#minimal-reproducible-examples---helping-others-help-you"><i class="fa fa-check"></i><b>25.1</b> Minimal reproducible examples - helping others help you</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">EHA Modeling &amp; Analytics Handbook</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="seqs_aln" class="section level1 hasAnchor" number="21">
<h1><span class="header-section-number">21</span> Sequences alignment<a href="21-seqs_aln.html#seqs_aln" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="heredity" class="section level2 hasAnchor" number="21.1">
<h2><span class="header-section-number">21.1</span> Heredity<a href="21-seqs_aln.html#heredity" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Today, we are capable of reconstructing kinship relationships between organisms through bioinformatics analysis of biological sequences, particularly proteins, and DNA. This is possible thanks to the principle of heredity. The biological sequences are transmitted from one generation to the next through sexual and asexual reproduction. From one generation to the next, the sequences accumulate variation as a result of stochastic or selective processes. Despite this variation, sequences from two closely related organisms are expected to be more similar to each other than sequences from two distantly related organisms. Between two organisms with a distant relationship, more time has elapsed since the last shared ancestor, accumulating more variation. This is the principle on which the search for homology between sequences is based, two DNA or protein sequences are homologous if they share ancestry. This means that the similarity between these sequences is the product of sharing the same ancestor placed somewhere in the phylogenetic tree of life. The higher the sequence similarity, the more recent the common ancestor between the sequences is expected to be. However, the identification of a hit with a high similarity is not enough to determine the homology of a sequence; it is also necessary to model the evolutionary process through phylogenetic tools. This chapter will address the methodological principles of pairwise alignments, the alignment between two biological sequences. Multiple sequence alignment is the topic of the next chapter. Before introducing pairwise alignment, I would like to introduce some homology concepts that are essential to the application and interpretation of alignments and phylogenies.</p>
</div>
<div id="orthologs-and-paralogs" class="section level2 hasAnchor" number="21.2">
<h2><span class="header-section-number">21.2</span> Orthologs and paralogs<a href="21-seqs_aln.html#orthologs-and-paralogs" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Homology relationships between sequences are defined by two important events in the evolution of species: duplication and speciation. Two sequences (genes or proteins) are orthologous if they are the product of speciation. Two sequences are paralogs if they are the products of duplication events. When two different organisms present sequences that are the product of duplication events prior to speciation, these sequences are out-paralogs. When duplication occurs after the speciation process, the sequences are called in-paralogs.</p>
</div>
<div id="blast" class="section level2 hasAnchor" number="21.3">
<h2><span class="header-section-number">21.3</span> BLAST<a href="21-seqs_aln.html#blast" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="conda-installation" class="section level3 hasAnchor" number="21.3.1">
<h3><span class="header-section-number">21.3.1</span> Conda Installation<a href="21-seqs_aln.html#conda-installation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>One of the easy ways to install bioinformatics software is with <a href="https://docs.conda.io/en/latest/">Conda</a>.
Download miniconda for ubuntu (If you have another OS choose the appropriate file)</p>
<pre><code>https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh</code></pre>
<p>Set the correct authorizations</p>
<pre><code>chmod +x Miniconda3-latest-Linux-x86_64.sh</code></pre>
<p>Install</p>
<pre><code>bash Miniconda3-latest-Linux-x86_64.sh
</code></pre>
<p>Export path</p>
<pre><code>export PATH=&quot;$HOME/miniconda3/bin:$PATH&quot;
</code></pre>
<p>Reinitialize the session.</p>
</div>
</div>
<div id="install-blast-with-conda" class="section level2 hasAnchor" number="21.4">
<h2><span class="header-section-number">21.4</span> Install <a href="https://www.ncbi.nlm.nih.gov/books/NBK279690/">BLAST+</a> with Conda<a href="21-seqs_aln.html#install-blast-with-conda" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<pre><code>conda install -c bioconda blast=2.12.0
</code></pre>
</div>
<div id="install-database-reference-sequences" class="section level2 hasAnchor" number="21.5">
<h2><span class="header-section-number">21.5</span> Install database reference sequences<a href="21-seqs_aln.html#install-database-reference-sequences" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The first step is to create the directories where the databases will be installed.
For example, in your home directory:</p>
<pre><code>
mkdir NCBI
</code></pre>
<pre><code>
cd NCBI
</code></pre>
<p>Create two subdirectories, one for nucleic acid sequences and one for protein sequences.</p>
<pre><code>
mkdir nt
</code></pre>
<pre><code>
mkdir nr
</code></pre>
<div id="there-are-different-ways-to-install-the-pre-formatted-blast-databases" class="section level3 hasAnchor" number="21.5.1">
<h3><span class="header-section-number">21.5.1</span> There are different ways to install the pre-formatted BLAST databases:<a href="21-seqs_aln.html#there-are-different-ways-to-install-the-pre-formatted-blast-databases" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="the-wget-command" class="section level4 hasAnchor" number="21.5.1.1">
<h4><span class="header-section-number">21.5.1.1</span> The ‘wget’ command<a href="21-seqs_aln.html#the-wget-command" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Download NCBI nucleotide sequence database</p>
<pre><code>
cd nt
</code></pre>
<pre><code>
wget ftp://ftp.ncbi.nlm.nih.gov/blast/db/nt.??.tar.gz
</code></pre>
<p>Uncompress the database files. You will find the uncompress.sh script in the scripts directory of this repository</p>
<pre><code>
./uncompress.sh
</code></pre>
<p>Download NCBI amino acid sequence database (In the nr directory)</p>
<pre><code>wget ftp://ftp.ncbi.nlm.nih.gov/blast/db/nr.??.tar.gz
</code></pre>
<pre><code>./uncompress.sh
</code></pre>
</div>
<div id="the-update_blastdb.pl-script." class="section level4 hasAnchor" number="21.5.1.2">
<h4><span class="header-section-number">21.5.1.2</span> The update_blastdb.pl script.<a href="21-seqs_aln.html#the-update_blastdb.pl-script." class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>BLAST+ offers the <a href="https://www.ncbi.nlm.nih.gov/books/NBK569850/">update_blastdb.pl</a> script that allows you to download the NCBI databases. This script uncompress the files for you.</p>
<p>Obtain information from the available databases to download</p>
<pre><code>
update_blastdb.pl --showall [*]
</code></pre>
<p>Download the nucleotide sequence database (NCBI_DB/nt directory)</p>
<pre><code>
update_blastdb.pl --decompress nt [*]
</code></pre>
<p>Download the amino acid database (NCBI_DB/nr directory)</p>
<pre><code>
update_blastdb.pl --decompress nr [*]
</code></pre>
</div>
</div>
</div>
<div id="pairwise-sequence-alignment" class="section level2 hasAnchor" number="21.6">
<h2><span class="header-section-number">21.6</span> Pairwise sequence alignment<a href="21-seqs_aln.html#pairwise-sequence-alignment" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="blastn" class="section level3 hasAnchor" number="21.6.1">
<h3><span class="header-section-number">21.6.1</span> BLASTn<a href="21-seqs_aln.html#blastn" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here is an example of alignment with BLASTn. Remember that BLAST offers different softwares to perform alignments according to the type of sequence and the reference database (<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST software</a>). BLASTn involves nucleotide-to-nucleotide alignments, this means that both the query sequences and the reference database are nucleotide sequences.</p>
<pre><code>
blastn -db /home/alexarmero2022/shared/NCBI/nt/nt   -evalue 0.0000000001     -num_threads 10  -outfmt &#39;6 std qcovs slen&#39; -query  test.fasta  &gt;    testout.csv

</code></pre>
<p>In the last command line there are several important points to keep in mind:</p>
<ol style="list-style-type: decimal">
<li><p>Specify the full directory path to the NCBI database. In this example I’m using my path (/home/alexarmero2022/shared/NCBI/nt), replace it with the appropriate directory path on your system.</p></li>
<li><p>The -db option expects the name of a database. It is for this reason that in addition to the path, it is also necessary to give the name of the database in the previous example nt.</p></li>
<li><p>The -num_threads option refers to the number of cores or logic CPUs that the user wants to assign to the task. Specify it according to availability in your system.</p></li>
<li><p>The results of an alignment can be reported in different types of formats (-outfmt). In the example, the tabular format (6) was used; this format is easy to manipulate. In this example the standard alignment information (std) was requested to be printed, but also some other information such as the alignment coverage in the query sequence (qcovs) and the length of the target sequence (slen). <a href="https://www.metagenomics.wiki/tools/blast/blastn-output-format-6">Tabular format options</a>.</p></li>
</ol>
</div>
<div id="blastp" class="section level3 hasAnchor" number="21.6.2">
<h3><span class="header-section-number">21.6.2</span> BLASTp<a href="21-seqs_aln.html#blastp" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here an example of protein-protein alignment with BLASTp. Notice that the path to the database changed; now it points to the nr directory, where the amino acid sequences were installed. In the same way the name of the database also changed (nr). The sequences in the test.fasta file must necessarily be amino acid sequences.</p>
<pre><code>
blastp -db /home/alexarmero2022/shared/NCBI/nr/nr   -evalue 0.0000000001     -num_threads 10  -outfmt &#39;6 std qcovs slen&#39; -query  test.fasta  &gt;    testout.csv

</code></pre>
</div>
<div id="configuration-files-and-environmental-variables" class="section level3 hasAnchor" number="21.6.3">
<h3><span class="header-section-number">21.6.3</span> Configuration files and environmental variables<a href="21-seqs_aln.html#configuration-files-and-environmental-variables" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>BLAST+ allows you to configure the blast runs. Configuration can be done through a .ncbirc file. In the scripts directory of this directory, there is an example of this file. Once the variables of interest have been modified, place this file in your home directory. For example, you could indicate where your NCBI database is located:</p>
<pre><code>
; Start the section for BLAST configuration
[BLAST]
; Specifies the path where BLAST databases are installed
BLASTDB=/home/alexarmero2022/shared/NCBI/
    
    </code></pre>
<p>In this last example we modify the variable <strong>BLASTDB</strong> and we assign the directory path to the NCBI databases. Note, that is the path we use in the blastn and blastp alignment examples. With this modification the blastn command line would subtly change:</p>
<pre><code>
blastn -db nt/nt   -evalue 0.0000000001     -num_threads 10  -outfmt &#39;6 std qcovs slen&#39; -query  test.fasta  &gt;    testout.csv

</code></pre>
<p>You could also use environmental variables. For example from your command line, you could write:</p>
<pre><code>BLASTDB=/home/alexarmero2022/shared/NCBI/
    </code></pre>
<p>There are several other <a href="https://www.ncbi.nlm.nih.gov/books/NBK569858/">variables</a> that can modify.</p>
<p>This last line is equivalent to using the .ncbirc file.</p>
<p>Note. The .ncbirc is a hidden file. To be able to observe it in your directory, execute the following line:</p>
<pre><code>ls -ad .*
    </code></pre>
</div>
</div>
<div id="blast-with-high-throughput-data" class="section level2 hasAnchor" number="21.7">
<h2><span class="header-section-number">21.7</span> BLAST with high-throughput data<a href="21-seqs_aln.html#blast-with-high-throughput-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>BLAST is perhaps the best software in the identification of homology relationships between biological sequences. The sensitivity of this tool has a cost both in the use of physical computing resources and in processing time. These limitations are particularly evident when are used NCBI databases. Several efforts have been made to use high performance computing (HPC) resources to efficiently run BLAST on large datasets. Here is one such tool: <a href="https://webchem.ncbr.muni.cz/Platform/App/CrocoBLAST">Crocroblast</a>. This software is presented as an example, and a starting point for higher performance strategies. Crocroblast determines what is the optimal file size to perform the alignment according to the HPC physical capabilities. In this way this software divides a large fasta file into files with sizes optimized to the available system resources. This is an easy to use tool; we recommend a good reading of the CrocroBLAST <a href="https://webchem.ncbr.muni.cz/Wiki/CrocoBLAST:UserManual">documentation</a>. The following example shows how to perform a CrocoBLAST blastn alignment.</p>
<div id="installation-2" class="section level3 hasAnchor" number="21.7.1">
<h3><span class="header-section-number">21.7.1</span> Installation<a href="21-seqs_aln.html#installation-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Get the right version for your <a href="https://webchem.ncbr.muni.cz/Platform/App/CrocoBLAST">operating system</a>.
Install the software in an appropriate directory</p>
<pre><code>mkdir Crocroblast
cd Crocroblast
unzip CrocoBLAST_linux_64.zip
chmod +x crocoblast
</code></pre>
</div>
<div id="add-sequence-database" class="section level3 hasAnchor" number="21.7.2">
<h3><span class="header-section-number">21.7.2</span> Add sequence database<a href="21-seqs_aln.html#add-sequence-database" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In order to use CrocroBLAST you need to add the sequence database.
Please adapt the following line according to the directory path where your database is located:</p>
<pre><code>./crocoblast -add_database --formatted_db /home/alexarmero2022/shared/NCBI/nt/nt.00.nsq
</code></pre>
</div>
<div id="add-the-job-to-the-queue" class="section level3 hasAnchor" number="21.7.3">
<h3><span class="header-section-number">21.7.3</span> Add the job to the queue<a href="21-seqs_aln.html#add-the-job-to-the-queue" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>CrocroBLAST handles job queues; this software has different functions to stop, pause, delete and run a job. The next line adds a job to the queue:</p>
<pre><code>./crocoblast -add_to_queue blastn nt data/test.fa output_test_4  --blast_options -outfmt 6
</code></pre>
<p>The following line will allow you to check if the job was added to the queue</p>
<pre><code>./crocoblast -status
</code></pre>
</div>
<div id="run-the-job" class="section level3 hasAnchor" number="21.7.4">
<h3><span class="header-section-number">21.7.4</span> Run the job<a href="21-seqs_aln.html#run-the-job" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<pre><code>./crocoblast -run --num_threads 4
</code></pre>
<p>CrocoBLAST shows messages with the time remaining to complete the process. We hope that this example motivates you to carefully read the CrocoBLAST <a href="https://webchem.ncbr.muni.cz/Wiki/CrocoBLAST:UserManual">wiki</a> to efficiently exploit the different possibilities of this nice tool.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="20-assembly-of-high-throughput-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="22-data-sources.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2,
"id": 1
},
"edit": {
"link": "https://github.com/ecohealthalliance/eha-ma-handbook/edit/master/Alignment.Rmd",
"text": "Edit this chapter"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section",
"scroll_highlight": true
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
